channeld/full_channel.c

```diff
@@ -443,8 +444,16 @@ static enum channel_add_err add_htlc(struct channel *channel,
 	 *     HTLCs to its local commitment transaction...
 	 *     - SHOULD fail the channel.
 	 */
+	/* Also we should not add more htlc's than sender or recipient
+	 * configured.  This mitigates attacks in which a peer can force the
+	 * funder of the channel to pay unnecessary onchain fees during a fee
+	 * spike with large commitment transactions.
+	 */
+	min_concurrent_htlcs = channel->config[recipient].max_accepted_htlcs;
+	if (min_concurrent_htlcs > channel->config[sender].max_accepted_htlcs)
+		min_concurrent_htlcs = channel->config[sender].max_accepted_htlcs;
 	if (tal_count(committed) - tal_count(removing) + tal_count(adding)
-	    > channel->config[recipient].max_accepted_htlcs) {
+	    > min_concurrent_htlcs) {
 		return CHANNEL_ERR_TOO_MANY_HTLCS;
```

Comment:
Seems to me that the spec specifies only counting htlc in one direction (not both) against this limit, further it only specifies comparing against the recipient's constraint.
Obviously you can enforce arbitrary restrictions for outgoing HTLCs, but if you enforce stricter restrictions than the spec says for incoming ones and then you force-close when (not if) they are violated by the remote then that's not ideal?

See what eclair does (as above) here:
https://github.com/ACINQ/eclair/blob/eae113f0988069044dc268d8a7f16728563baa5f/eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala#L216
https://github.com/ACINQ/eclair/blob/eae113f0988069044dc268d8a7f16728563baa5f/eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala#L259
(check only one direction of HTLCs against only one of the constraints)

Reply:
Oh, I've just realised that the channel does not actually get force-closed, only the peer get disconnected from (and then the htlc is not added as it was not committed).
BOLT-02 specifically says "fail the channel" - I guess I had a different interpretation of that...
I mean BOLT-02 sometimes says "fail the connection" and sometimes "fail the channel", but then that means c-lightning considers those both to be identical?
Re the original remark, I guess it's less bad then, but still not ideal.

Reply:
Note: my comments have been addressed in https://github.com/ElementsProject/lightning/pull/4432 which reverted the strict checks for remote behaviour

Reply:
Actually, only one of the two concerns have been addressed by #4432.
AFAICT c-lightning is still checking the total number of htlcs (sum of both directions) against `max_accepted_htlcs`, while I think BOLT-02 only specifies to check the number of htlcs the sender offered (so one direction only).
This is means c-lightning is still stricter than e.g. eclair.
https://github.com/ElementsProject/lightning/blob/9470ea3032a00438504e3769a65212fe22165717/channeld/full_channel.c#L575-L578
Or am I misunderstanding the code here? It seems to me `committed`, `removing`, and `adding` include htlcs in both directions.
@rustyrussell